upstream-sha256: e14c1b857431a8f9aa045f31c10599a90b255f6e994a7110dcc80f030eb676b1
actual-sha256: cc201e25c2ac94fda63a8b9d4de00a1248967d8f88b77d508989637b89f968d5
--- upstream.expected
+++ tolk.actual
@@ -11,55 +11,58 @@
    9 Ops.SHL             :            dtypes.int                               [6, 8]                           None
   10 Ops.ADD             :            dtypes.int                               [3, 9]                           None
   11 Ops.INDEX           :            dtypes.float                             [0, 10]                          None
-  12 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 512, device=('CPU:0', 'CPU:1', 'CPU:2', 'CPU:3'))
-  13 Ops.INDEX           :            dtypes.float                             [12, 10]                         None
-  14 Ops.LOAD            :            dtypes.float                             [13]                             None
-  15 Ops.CONST           :            dtypes.weakint                           []                               64
-  16 Ops.CAST            :            dtypes.int                               ['64']                           dtypes.int
-  17 Ops.ADD             :            dtypes.int                               [10, 16]                         None
-  18 Ops.INDEX           :            dtypes.float                             [12, 17]                         None
-  19 Ops.LOAD            :            dtypes.float                             [18]                             None
-  20 Ops.ADD             :            dtypes.float                             [14, 19]                         None
-  21 Ops.CONST           :            dtypes.weakint                           []                               128
-  22 Ops.CAST            :            dtypes.int                               ['128']                          dtypes.int
-  23 Ops.ADD             :            dtypes.int                               [10, 22]                         None
-  24 Ops.INDEX           :            dtypes.float                             [12, 23]                         None
-  25 Ops.LOAD            :            dtypes.float                             [24]                             None
-  26 Ops.ADD             :            dtypes.float                             [20, 25]                         None
-  27 Ops.CONST           :            dtypes.weakint                           []                               192
-  28 Ops.CAST            :            dtypes.int                               ['192']                          dtypes.int
-  29 Ops.ADD             :            dtypes.int                               [10, 28]                         None
-  30 Ops.INDEX           :            dtypes.float                             [12, 29]                         None
-  31 Ops.LOAD            :            dtypes.float                             [30]                             None
-  32 Ops.ADD             :            dtypes.float                             [26, 31]                         None
-  33 Ops.CONST           :            dtypes.weakint                           []                               256
-  34 Ops.CAST            :            dtypes.int                               ['256']                          dtypes.int
-  35 Ops.ADD             :            dtypes.int                               [10, 34]                         None
-  36 Ops.INDEX           :            dtypes.float                             [12, 35]                         None
-  37 Ops.LOAD            :            dtypes.float                             [36]                             None
-  38 Ops.ADD             :            dtypes.float                             [32, 37]                         None
-  39 Ops.CONST           :            dtypes.weakint                           []                               320
-  40 Ops.CAST            :            dtypes.int                               ['320']                          dtypes.int
-  41 Ops.ADD             :            dtypes.int                               [10, 40]                         None
-  42 Ops.INDEX           :            dtypes.float                             [12, 41]                         None
-  43 Ops.LOAD            :            dtypes.float                             [42]                             None
-  44 Ops.ADD             :            dtypes.float                             [38, 43]                         None
-  45 Ops.CONST           :            dtypes.weakint                           []                               384
-  46 Ops.CAST            :            dtypes.int                               ['384']                          dtypes.int
-  47 Ops.ADD             :            dtypes.int                               [10, 46]                         None
-  48 Ops.INDEX           :            dtypes.float                             [12, 47]                         None
-  49 Ops.LOAD            :            dtypes.float                             [48]                             None
-  50 Ops.ADD             :            dtypes.float                             [44, 49]                         None
-  51 Ops.CONST           :            dtypes.weakint                           []                               448
-  52 Ops.CAST            :            dtypes.int                               ['448']                          dtypes.int
-  53 Ops.ADD             :            dtypes.int                               [10, 52]                         None
-  54 Ops.INDEX           :            dtypes.float                             [12, 53]                         None
-  55 Ops.LOAD            :            dtypes.float                             [54]                             None
-  56 Ops.ADD             :            dtypes.float                             [50, 55]                         None
-  57 Ops.STORE           :            dtypes.void                              [11, 56]                         None
-  58 Ops.SINK            :            dtypes.void                              [57]                             KernelInfo(name='r_2_32_8', applied_opts=(Opt(op=OptOps.SPLIT, axis=1, arg=(0, AxisType.UNROLL)), Opt(op=OptOps.SPLIT, axis=0, arg=(32, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
+  12 Ops.CONST           :            dtypes.weakfloat                         []                               0.0
+  13 Ops.CAST            :            dtypes.float                             ['0.0']                          dtypes.float
+  14 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 512, device=('CPU:0', 'CPU:1', 'CPU:2', 'CPU:3'))
+  15 Ops.INDEX           :            dtypes.float                             [14, 10]                         None
+  16 Ops.LOAD            :            dtypes.float                             [15]                             None
+  17 Ops.ADD             :            dtypes.float                             [13, 16]                         None
+  18 Ops.CONST           :            dtypes.weakint                           []                               64
+  19 Ops.CAST            :            dtypes.int                               ['64']                           dtypes.int
+  20 Ops.ADD             :            dtypes.int                               [10, 19]                         None
+  21 Ops.INDEX           :            dtypes.float                             [14, 20]                         None
+  22 Ops.LOAD            :            dtypes.float                             [21]                             None
+  23 Ops.ADD             :            dtypes.float                             [17, 22]                         None
+  24 Ops.CONST           :            dtypes.weakint                           []                               128
+  25 Ops.CAST            :            dtypes.int                               ['128']                          dtypes.int
+  26 Ops.ADD             :            dtypes.int                               [10, 25]                         None
+  27 Ops.INDEX           :            dtypes.float                             [14, 26]                         None
+  28 Ops.LOAD            :            dtypes.float                             [27]                             None
+  29 Ops.ADD             :            dtypes.float                             [23, 28]                         None
+  30 Ops.CONST           :            dtypes.weakint                           []                               192
+  31 Ops.CAST            :            dtypes.int                               ['192']                          dtypes.int
+  32 Ops.ADD             :            dtypes.int                               [10, 31]                         None
+  33 Ops.INDEX           :            dtypes.float                             [14, 32]                         None
+  34 Ops.LOAD            :            dtypes.float                             [33]                             None
+  35 Ops.ADD             :            dtypes.float                             [29, 34]                         None
+  36 Ops.CONST           :            dtypes.weakint                           []                               256
+  37 Ops.CAST            :            dtypes.int                               ['256']                          dtypes.int
+  38 Ops.ADD             :            dtypes.int                               [10, 37]                         None
+  39 Ops.INDEX           :            dtypes.float                             [14, 38]                         None
+  40 Ops.LOAD            :            dtypes.float                             [39]                             None
+  41 Ops.ADD             :            dtypes.float                             [35, 40]                         None
+  42 Ops.CONST           :            dtypes.weakint                           []                               320
+  43 Ops.CAST            :            dtypes.int                               ['320']                          dtypes.int
+  44 Ops.ADD             :            dtypes.int                               [10, 43]                         None
+  45 Ops.INDEX           :            dtypes.float                             [14, 44]                         None
+  46 Ops.LOAD            :            dtypes.float                             [45]                             None
+  47 Ops.ADD             :            dtypes.float                             [41, 46]                         None
+  48 Ops.CONST           :            dtypes.weakint                           []                               384
+  49 Ops.CAST            :            dtypes.int                               ['384']                          dtypes.int
+  50 Ops.ADD             :            dtypes.int                               [10, 49]                         None
+  51 Ops.INDEX           :            dtypes.float                             [14, 50]                         None
+  52 Ops.LOAD            :            dtypes.float                             [51]                             None
+  53 Ops.ADD             :            dtypes.float                             [47, 52]                         None
+  54 Ops.CONST           :            dtypes.weakint                           []                               448
+  55 Ops.CAST            :            dtypes.int                               ['448']                          dtypes.int
+  56 Ops.ADD             :            dtypes.int                               [10, 55]                         None
+  57 Ops.INDEX           :            dtypes.float                             [14, 56]                         None
+  58 Ops.LOAD            :            dtypes.float                             [57]                             None
+  59 Ops.ADD             :            dtypes.float                             [53, 58]                         None
+  60 Ops.STORE           :            dtypes.void                              [11, 59]                         None
+  61 Ops.SINK            :            dtypes.void                              [60]                             KernelInfo(name='r_2_32_8', applied_opts=(Opt(op=OptOps.SPLIT, axis=1, arg=(0, AxisType.UNROLL)), Opt(op=OptOps.SPLIT, axis=0, arg=(32, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
 === kernel 1 ===
-   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:0')
+   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:1')
    1 Ops.CONST           :            dtypes.weakint                           []                               8
    2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
    3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
@@ -76,23 +79,6 @@
   14 Ops.GROUP           :            dtypes.void                              [13, 13, 13, 13]                 None
   15 Ops.SINK            :            dtypes.void                              [14]                             KernelInfo(name='E_8_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(4, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=0, arg=(8, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
 === kernel 2 ===
-   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:1')
-   1 Ops.CONST           :            dtypes.weakint                           []                               8
-   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
-   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
-   4 Ops.CONST           :            dtypes.weakint                           []                               2
-   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
-   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
-   7 Ops.CONST           :            dtypes.weakint                           []                               4
-   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
-   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
-  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:0')
-  11 Ops.SHRINK          :            dtypes.float                             [10, 6, 8]                       None
-  12 Ops.LOAD            :            dtypes.float                             [11]                             None
-  13 Ops.STORE           :            dtypes.void                              [9, 12]                          None
-  14 Ops.GROUP           :            dtypes.void                              [13, 13, 13, 13]                 None
-  15 Ops.SINK            :            dtypes.void                              [14]                             KernelInfo(name='E_8_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(4, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=0, arg=(8, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
-=== kernel 3 ===
    0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:1')
    1 Ops.CONST           :            dtypes.weakint                           []                               8
    2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
@@ -131,24 +117,24 @@
   35 Ops.STORE           :            dtypes.void                              [9, 34]                          None
   36 Ops.GROUP           :            dtypes.void                              [35, 35, 35, 35]                 None
   37 Ops.SINK            :            dtypes.void                              [36]                             KernelInfo(name='E_8_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(4, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=0, arg=(8, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
+=== kernel 3 ===
+   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:2')
+   1 Ops.CONST           :            dtypes.weakint                           []                               8
+   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
+   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
+   4 Ops.CONST           :            dtypes.weakint                           []                               2
+   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
+   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
+   7 Ops.CONST           :            dtypes.weakint                           []                               4
+   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
+   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
+  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:1')
+  11 Ops.SHRINK          :            dtypes.float                             [10, 6, 8]                       None
+  12 Ops.LOAD            :            dtypes.float                             [11]                             None
+  13 Ops.STORE           :            dtypes.void                              [9, 12]                          None
+  14 Ops.GROUP           :            dtypes.void                              [13, 13, 13, 13]                 None
+  15 Ops.SINK            :            dtypes.void                              [14]                             KernelInfo(name='E_8_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(4, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=0, arg=(8, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
 === kernel 4 ===
-   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:2')
-   1 Ops.CONST           :            dtypes.weakint                           []                               8
-   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
-   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
-   4 Ops.CONST           :            dtypes.weakint                           []                               2
-   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
-   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
-   7 Ops.CONST           :            dtypes.weakint                           []                               4
-   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
-   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
-  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:1')
-  11 Ops.SHRINK          :            dtypes.float                             [10, 6, 8]                       None
-  12 Ops.LOAD            :            dtypes.float                             [11]                             None
-  13 Ops.STORE           :            dtypes.void                              [9, 12]                          None
-  14 Ops.GROUP           :            dtypes.void                              [13, 13, 13, 13]                 None
-  15 Ops.SINK            :            dtypes.void                              [14]                             KernelInfo(name='E_8_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(4, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=0, arg=(8, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
-=== kernel 5 ===
    0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:2')
    1 Ops.CONST           :            dtypes.weakint                           []                               8
    2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
@@ -187,24 +173,24 @@
   35 Ops.STORE           :            dtypes.void                              [9, 34]                          None
   36 Ops.GROUP           :            dtypes.void                              [35, 35, 35, 35]                 None
   37 Ops.SINK            :            dtypes.void                              [36]                             KernelInfo(name='E_8_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(4, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=0, arg=(8, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
+=== kernel 5 ===
+   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:3')
+   1 Ops.CONST           :            dtypes.weakint                           []                               8
+   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
+   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
+   4 Ops.CONST           :            dtypes.weakint                           []                               2
+   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
+   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
+   7 Ops.CONST           :            dtypes.weakint                           []                               4
+   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
+   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
+  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:2')
+  11 Ops.SHRINK          :            dtypes.float                             [10, 6, 8]                       None
+  12 Ops.LOAD            :            dtypes.float                             [11]                             None
+  13 Ops.STORE           :            dtypes.void                              [9, 12]                          None
+  14 Ops.GROUP           :            dtypes.void                              [13, 13, 13, 13]                 None
+  15 Ops.SINK            :            dtypes.void                              [14]                             KernelInfo(name='E_8_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(4, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=0, arg=(8, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
 === kernel 6 ===
-   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:3')
-   1 Ops.CONST           :            dtypes.weakint                           []                               8
-   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
-   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
-   4 Ops.CONST           :            dtypes.weakint                           []                               2
-   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
-   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
-   7 Ops.CONST           :            dtypes.weakint                           []                               4
-   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
-   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
-  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:2')
-  11 Ops.SHRINK          :            dtypes.float                             [10, 6, 8]                       None
-  12 Ops.LOAD            :            dtypes.float                             [11]                             None
-  13 Ops.STORE           :            dtypes.void                              [9, 12]                          None
-  14 Ops.GROUP           :            dtypes.void                              [13, 13, 13, 13]                 None
-  15 Ops.SINK            :            dtypes.void                              [14]                             KernelInfo(name='E_8_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(4, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=0, arg=(8, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
-=== kernel 7 ===
    0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:3')
    1 Ops.CONST           :            dtypes.weakint                           []                               8
    2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
@@ -243,58 +229,58 @@
   35 Ops.STORE           :            dtypes.void                              [9, 34]                          None
   36 Ops.GROUP           :            dtypes.void                              [35, 35, 35, 35]                 None
   37 Ops.SINK            :            dtypes.void                              [36]                             KernelInfo(name='E_8_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(4, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=0, arg=(8, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
+=== kernel 7 ===
+   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:0')
+   1 Ops.CONST           :            dtypes.weakint                           []                               8
+   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
+   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
+   4 Ops.CONST           :            dtypes.weakint                           []                               2
+   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
+   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
+   7 Ops.CONST           :            dtypes.weakint                           []                               4
+   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
+   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
+  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:3')
+  11 Ops.SHRINK          :            dtypes.float                             [10, 6, 8]                       None
+  12 Ops.LOAD            :            dtypes.float                             [11]                             None
+  13 Ops.STORE           :            dtypes.void                              [9, 12]                          None
+  14 Ops.GROUP           :            dtypes.void                              [13, 13, 13, 13]                 None
+  15 Ops.SINK            :            dtypes.void                              [14]                             KernelInfo(name='E_8_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(4, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=0, arg=(8, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
 === kernel 8 ===
-   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:0')
-   1 Ops.CONST           :            dtypes.weakint                           []                               8
-   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
-   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
-   4 Ops.CONST           :            dtypes.weakint                           []                               2
-   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
-   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
-   7 Ops.CONST           :            dtypes.weakint                           []                               4
-   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
-   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
-  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:3')
+   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:1')
+   1 Ops.CONST           :            dtypes.weakint                           []                               8
+   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
+   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
+   4 Ops.CONST           :            dtypes.weakint                           []                               2
+   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
+   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
+   7 Ops.CONST           :            dtypes.weakint                           []                               4
+   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
+   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
+  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:0')
   11 Ops.SHRINK          :            dtypes.float                             [10, 6, 8]                       None
   12 Ops.LOAD            :            dtypes.float                             [11]                             None
   13 Ops.STORE           :            dtypes.void                              [9, 12]                          None
   14 Ops.GROUP           :            dtypes.void                              [13, 13, 13, 13]                 None
   15 Ops.SINK            :            dtypes.void                              [14]                             KernelInfo(name='E_8_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(4, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=0, arg=(8, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
 === kernel 9 ===
-   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:1')
-   1 Ops.CONST           :            dtypes.weakint                           []                               8
-   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
-   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
-   4 Ops.CONST           :            dtypes.weakint                           []                               2
-   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
-   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
-   7 Ops.CONST           :            dtypes.weakint                           []                               4
-   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
-   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
-  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:0')
+   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:2')
+   1 Ops.CONST           :            dtypes.weakint                           []                               8
+   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
+   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
+   4 Ops.CONST           :            dtypes.weakint                           []                               2
+   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
+   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
+   7 Ops.CONST           :            dtypes.weakint                           []                               4
+   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
+   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
+  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:1')
   11 Ops.SHRINK          :            dtypes.float                             [10, 6, 8]                       None
   12 Ops.LOAD            :            dtypes.float                             [11]                             None
   13 Ops.STORE           :            dtypes.void                              [9, 12]                          None
   14 Ops.GROUP           :            dtypes.void                              [13, 13, 13, 13]                 None
   15 Ops.SINK            :            dtypes.void                              [14]                             KernelInfo(name='E_8_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(4, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=0, arg=(8, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
 === kernel 10 ===
-   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:2')
-   1 Ops.CONST           :            dtypes.weakint                           []                               8
-   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
-   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
-   4 Ops.CONST           :            dtypes.weakint                           []                               2
-   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
-   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
-   7 Ops.CONST           :            dtypes.weakint                           []                               4
-   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
-   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
-  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:1')
-  11 Ops.SHRINK          :            dtypes.float                             [10, 6, 8]                       None
-  12 Ops.LOAD            :            dtypes.float                             [11]                             None
-  13 Ops.STORE           :            dtypes.void                              [9, 12]                          None
-  14 Ops.GROUP           :            dtypes.void                              [13, 13, 13, 13]                 None
-  15 Ops.SINK            :            dtypes.void                              [14]                             KernelInfo(name='E_8_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(4, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=0, arg=(8, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
-=== kernel 11 ===
    0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:3')
    1 Ops.CONST           :            dtypes.weakint                           []                               8
    2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
@@ -333,8 +319,8 @@
   35 Ops.STORE           :            dtypes.void                              [9, 34]                          None
   36 Ops.GROUP           :            dtypes.void                              [35, 35, 35, 35]                 None
   37 Ops.SINK            :            dtypes.void                              [36]                             KernelInfo(name='E_8_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(4, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=0, arg=(8, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
-=== kernel 12 ===
-   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:1')
+=== kernel 11 ===
+   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:2')
    1 Ops.CONST           :            dtypes.weakint                           []                               8
    2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
    3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
@@ -353,41 +339,83 @@
   16 Ops.STORE           :            dtypes.void                              [9, 15]                          None
   17 Ops.GROUP           :            dtypes.void                              [16, 16, 16, 16]                 None
   18 Ops.SINK            :            dtypes.void                              [17]                             KernelInfo(name='E_8_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(4, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=0, arg=(8, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
+=== kernel 12 ===
+   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:2')
+   1 Ops.CONST           :            dtypes.weakint                           []                               8
+   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
+   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
+   4 Ops.CONST           :            dtypes.weakint                           []                               2
+   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
+   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
+   7 Ops.CONST           :            dtypes.weakint                           []                               4
+   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
+   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
+  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:2')
+  11 Ops.SHRINK          :            dtypes.float                             [10, 6, 8]                       None
+  12 Ops.LOAD            :            dtypes.float                             [11]                             None
+  13 Ops.CONST           :            dtypes.weakint                           []                               0
+  14 Ops.CAST            :            dtypes.int                               ['0']                            dtypes.int
+  15 Ops.INDEX           :            dtypes.float                             [12, 14]                         None
+  16 Ops.PARAM           :            dtypes.float                             []                               ParamArg(2, dtypes.float, 64, device='CPU:2')
+  17 Ops.CONST           :            dtypes.weakint                           []                               32
+  18 Ops.CAST            :            dtypes.int                               ['32']                           dtypes.int
+  19 Ops.ADD             :            dtypes.int                               [6, 18]                          None
+  20 Ops.SHRINK          :            dtypes.float                             [16, 19, 8]                      None
+  21 Ops.LOAD            :            dtypes.float                             [20]                             None
+  22 Ops.INDEX           :            dtypes.float                             [21, 14]                         None
+  23 Ops.ADD             :            dtypes.float                             [15, 22]                         None
+  24 Ops.CONST           :            dtypes.weakint                           []                               1
+  25 Ops.CAST            :            dtypes.int                               ['1']                            dtypes.int
+  26 Ops.INDEX           :            dtypes.float                             [12, 25]                         None
+  27 Ops.INDEX           :            dtypes.float                             [21, 25]                         None
+  28 Ops.ADD             :            dtypes.float                             [26, 27]                         None
+  29 Ops.INDEX           :            dtypes.float                             [12, 5]                          None
+  30 Ops.INDEX           :            dtypes.float                             [21, 5]                          None
+  31 Ops.ADD             :            dtypes.float                             [29, 30]                         None
+  32 Ops.CONST           :            dtypes.weakint                           []                               3
+  33 Ops.CAST            :            dtypes.int                               ['3']                            dtypes.int
+  34 Ops.INDEX           :            dtypes.float                             [12, 33]                         None
+  35 Ops.INDEX           :            dtypes.float                             [21, 33]                         None
+  36 Ops.ADD             :            dtypes.float                             [34, 35]                         None
+  37 Ops.STACK           :            dtypes.float                             [23, 28, 31, 36]                 None
+  38 Ops.STORE           :            dtypes.void                              [9, 37]                          None
+  39 Ops.GROUP           :            dtypes.void                              [38, 38, 38, 38]                 None
+  40 Ops.SINK            :            dtypes.void                              [39]                             KernelInfo(name='E_8_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(4, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=0, arg=(8, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
 === kernel 13 ===
-   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:2')
-   1 Ops.CONST           :            dtypes.weakint                           []                               8
-   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
-   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
-   4 Ops.CONST           :            dtypes.weakint                           []                               2
-   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
-   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
-   7 Ops.CONST           :            dtypes.weakint                           []                               4
-   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
-   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
-  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:1')
+   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:3')
+   1 Ops.CONST           :            dtypes.weakint                           []                               8
+   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
+   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
+   4 Ops.CONST           :            dtypes.weakint                           []                               2
+   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
+   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
+   7 Ops.CONST           :            dtypes.weakint                           []                               4
+   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
+   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
+  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:2')
   11 Ops.SHRINK          :            dtypes.float                             [10, 6, 8]                       None
   12 Ops.LOAD            :            dtypes.float                             [11]                             None
   13 Ops.STORE           :            dtypes.void                              [9, 12]                          None
   14 Ops.GROUP           :            dtypes.void                              [13, 13, 13, 13]                 None
   15 Ops.SINK            :            dtypes.void                              [14]                             KernelInfo(name='E_8_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(4, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=0, arg=(8, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
 === kernel 14 ===
-   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:2')
-   1 Ops.CONST           :            dtypes.weakint                           []                               8
-   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
-   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
-   4 Ops.CONST           :            dtypes.weakint                           []                               2
-   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
-   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
-   7 Ops.CONST           :            dtypes.weakint                           []                               4
-   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
-   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
-  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:2')
-  11 Ops.SHRINK          :            dtypes.float                             [10, 6, 8]                       None
-  12 Ops.LOAD            :            dtypes.float                             [11]                             None
-  13 Ops.CONST           :            dtypes.weakint                           []                               0
-  14 Ops.CAST            :            dtypes.int                               ['0']                            dtypes.int
-  15 Ops.INDEX           :            dtypes.float                             [12, 14]                         None
-  16 Ops.PARAM           :            dtypes.float                             []                               ParamArg(2, dtypes.float, 64, device='CPU:2')
+   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:3')
+   1 Ops.CONST           :            dtypes.weakint                           []                               8
+   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
+   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
+   4 Ops.CONST           :            dtypes.weakint                           []                               2
+   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
+   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
+   7 Ops.CONST           :            dtypes.weakint                           []                               4
+   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
+   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
+  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:3')
+  11 Ops.SHRINK          :            dtypes.float                             [10, 6, 8]                       None
+  12 Ops.LOAD            :            dtypes.float                             [11]                             None
+  13 Ops.CONST           :            dtypes.weakint                           []                               0
+  14 Ops.CAST            :            dtypes.int                               ['0']                            dtypes.int
+  15 Ops.INDEX           :            dtypes.float                             [12, 14]                         None
+  16 Ops.PARAM           :            dtypes.float                             []                               ParamArg(2, dtypes.float, 64, device='CPU:3')
   17 Ops.CONST           :            dtypes.weakint                           []                               32
   18 Ops.CAST            :            dtypes.int                               ['32']                           dtypes.int
   19 Ops.ADD             :            dtypes.int                               [6, 18]                          None
@@ -413,40 +441,40 @@
   39 Ops.GROUP           :            dtypes.void                              [38, 38, 38, 38]                 None
   40 Ops.SINK            :            dtypes.void                              [39]                             KernelInfo(name='E_8_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(4, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=0, arg=(8, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
 === kernel 15 ===
-   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:3')
-   1 Ops.CONST           :            dtypes.weakint                           []                               8
-   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
-   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
-   4 Ops.CONST           :            dtypes.weakint                           []                               2
-   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
-   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
-   7 Ops.CONST           :            dtypes.weakint                           []                               4
-   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
-   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
-  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:2')
+   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:0')
+   1 Ops.CONST           :            dtypes.weakint                           []                               8
+   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
+   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
+   4 Ops.CONST           :            dtypes.weakint                           []                               2
+   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
+   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
+   7 Ops.CONST           :            dtypes.weakint                           []                               4
+   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
+   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
+  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:3')
   11 Ops.SHRINK          :            dtypes.float                             [10, 6, 8]                       None
   12 Ops.LOAD            :            dtypes.float                             [11]                             None
   13 Ops.STORE           :            dtypes.void                              [9, 12]                          None
   14 Ops.GROUP           :            dtypes.void                              [13, 13, 13, 13]                 None
   15 Ops.SINK            :            dtypes.void                              [14]                             KernelInfo(name='E_8_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(4, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=0, arg=(8, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
 === kernel 16 ===
-   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:3')
-   1 Ops.CONST           :            dtypes.weakint                           []                               8
-   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
-   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
-   4 Ops.CONST           :            dtypes.weakint                           []                               2
-   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
-   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
-   7 Ops.CONST           :            dtypes.weakint                           []                               4
-   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
-   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
-  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:3')
-  11 Ops.SHRINK          :            dtypes.float                             [10, 6, 8]                       None
-  12 Ops.LOAD            :            dtypes.float                             [11]                             None
-  13 Ops.CONST           :            dtypes.weakint                           []                               0
-  14 Ops.CAST            :            dtypes.int                               ['0']                            dtypes.int
-  15 Ops.INDEX           :            dtypes.float                             [12, 14]                         None
-  16 Ops.PARAM           :            dtypes.float                             []                               ParamArg(2, dtypes.float, 64, device='CPU:3')
+   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:0')
+   1 Ops.CONST           :            dtypes.weakint                           []                               8
+   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
+   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
+   4 Ops.CONST           :            dtypes.weakint                           []                               2
+   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
+   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
+   7 Ops.CONST           :            dtypes.weakint                           []                               4
+   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
+   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
+  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:0')
+  11 Ops.SHRINK          :            dtypes.float                             [10, 6, 8]                       None
+  12 Ops.LOAD            :            dtypes.float                             [11]                             None
+  13 Ops.CONST           :            dtypes.weakint                           []                               0
+  14 Ops.CAST            :            dtypes.int                               ['0']                            dtypes.int
+  15 Ops.INDEX           :            dtypes.float                             [12, 14]                         None
+  16 Ops.PARAM           :            dtypes.float                             []                               ParamArg(2, dtypes.float, 64, device='CPU:0')
   17 Ops.CONST           :            dtypes.weakint                           []                               32
   18 Ops.CAST            :            dtypes.int                               ['32']                           dtypes.int
   19 Ops.ADD             :            dtypes.int                               [6, 18]                          None
@@ -482,23 +510,6 @@
    7 Ops.CONST           :            dtypes.weakint                           []                               4
    8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
    9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
-  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:3')
-  11 Ops.SHRINK          :            dtypes.float                             [10, 6, 8]                       None
-  12 Ops.LOAD            :            dtypes.float                             [11]                             None
-  13 Ops.STORE           :            dtypes.void                              [9, 12]                          None
-  14 Ops.GROUP           :            dtypes.void                              [13, 13, 13, 13]                 None
-  15 Ops.SINK            :            dtypes.void                              [14]                             KernelInfo(name='E_8_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(4, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=0, arg=(8, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
-=== kernel 18 ===
-   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:0')
-   1 Ops.CONST           :            dtypes.weakint                           []                               8
-   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
-   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
-   4 Ops.CONST           :            dtypes.weakint                           []                               2
-   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
-   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
-   7 Ops.CONST           :            dtypes.weakint                           []                               4
-   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
-   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
   10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:0')
   11 Ops.SHRINK          :            dtypes.float                             [10, 6, 8]                       None
   12 Ops.LOAD            :            dtypes.float                             [11]                             None
@@ -530,100 +541,58 @@
   38 Ops.STORE           :            dtypes.void                              [9, 37]                          None
   39 Ops.GROUP           :            dtypes.void                              [38, 38, 38, 38]                 None
   40 Ops.SINK            :            dtypes.void                              [39]                             KernelInfo(name='E_8_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(4, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=0, arg=(8, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
+=== kernel 18 ===
+   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:1')
+   1 Ops.CONST           :            dtypes.weakint                           []                               8
+   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
+   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
+   4 Ops.CONST           :            dtypes.weakint                           []                               2
+   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
+   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
+   7 Ops.CONST           :            dtypes.weakint                           []                               4
+   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
+   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
+  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:0')
+  11 Ops.SHRINK          :            dtypes.float                             [10, 6, 8]                       None
+  12 Ops.LOAD            :            dtypes.float                             [11]                             None
+  13 Ops.STORE           :            dtypes.void                              [9, 12]                          None
+  14 Ops.GROUP           :            dtypes.void                              [13, 13, 13, 13]                 None
+  15 Ops.SINK            :            dtypes.void                              [14]                             KernelInfo(name='E_8_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(4, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=0, arg=(8, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
 === kernel 19 ===
-   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:0')
-   1 Ops.CONST           :            dtypes.weakint                           []                               8
-   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
-   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
-   4 Ops.CONST           :            dtypes.weakint                           []                               2
-   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
-   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
-   7 Ops.CONST           :            dtypes.weakint                           []                               4
-   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
-   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
-  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:0')
-  11 Ops.SHRINK          :            dtypes.float                             [10, 6, 8]                       None
-  12 Ops.LOAD            :            dtypes.float                             [11]                             None
-  13 Ops.CONST           :            dtypes.weakint                           []                               0
-  14 Ops.CAST            :            dtypes.int                               ['0']                            dtypes.int
-  15 Ops.INDEX           :            dtypes.float                             [12, 14]                         None
-  16 Ops.PARAM           :            dtypes.float                             []                               ParamArg(2, dtypes.float, 64, device='CPU:0')
-  17 Ops.CONST           :            dtypes.weakint                           []                               32
-  18 Ops.CAST            :            dtypes.int                               ['32']                           dtypes.int
-  19 Ops.ADD             :            dtypes.int                               [6, 18]                          None
-  20 Ops.SHRINK          :            dtypes.float                             [16, 19, 8]                      None
-  21 Ops.LOAD            :            dtypes.float                             [20]                             None
-  22 Ops.INDEX           :            dtypes.float                             [21, 14]                         None
-  23 Ops.ADD             :            dtypes.float                             [15, 22]                         None
-  24 Ops.CONST           :            dtypes.weakint                           []                               1
-  25 Ops.CAST            :            dtypes.int                               ['1']                            dtypes.int
-  26 Ops.INDEX           :            dtypes.float                             [12, 25]                         None
-  27 Ops.INDEX           :            dtypes.float                             [21, 25]                         None
-  28 Ops.ADD             :            dtypes.float                             [26, 27]                         None
-  29 Ops.INDEX           :            dtypes.float                             [12, 5]                          None
-  30 Ops.INDEX           :            dtypes.float                             [21, 5]                          None
-  31 Ops.ADD             :            dtypes.float                             [29, 30]                         None
-  32 Ops.CONST           :            dtypes.weakint                           []                               3
-  33 Ops.CAST            :            dtypes.int                               ['3']                            dtypes.int
-  34 Ops.INDEX           :            dtypes.float                             [12, 33]                         None
-  35 Ops.INDEX           :            dtypes.float                             [21, 33]                         None
-  36 Ops.ADD             :            dtypes.float                             [34, 35]                         None
-  37 Ops.STACK           :            dtypes.float                             [23, 28, 31, 36]                 None
-  38 Ops.STORE           :            dtypes.void                              [9, 37]                          None
-  39 Ops.GROUP           :            dtypes.void                              [38, 38, 38, 38]                 None
-  40 Ops.SINK            :            dtypes.void                              [39]                             KernelInfo(name='E_8_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(4, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=0, arg=(8, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
+   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:2')
+   1 Ops.CONST           :            dtypes.weakint                           []                               8
+   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
+   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
+   4 Ops.CONST           :            dtypes.weakint                           []                               2
+   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
+   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
+   7 Ops.CONST           :            dtypes.weakint                           []                               4
+   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
+   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
+  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:1')
+  11 Ops.SHRINK          :            dtypes.float                             [10, 6, 8]                       None
+  12 Ops.LOAD            :            dtypes.float                             [11]                             None
+  13 Ops.STORE           :            dtypes.void                              [9, 12]                          None
+  14 Ops.GROUP           :            dtypes.void                              [13, 13, 13, 13]                 None
+  15 Ops.SINK            :            dtypes.void                              [14]                             KernelInfo(name='E_8_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(4, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=0, arg=(8, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
 === kernel 20 ===
-   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:1')
-   1 Ops.CONST           :            dtypes.weakint                           []                               8
-   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
-   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
-   4 Ops.CONST           :            dtypes.weakint                           []                               2
-   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
-   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
-   7 Ops.CONST           :            dtypes.weakint                           []                               4
-   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
-   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
-  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:0')
+   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:3')
+   1 Ops.CONST           :            dtypes.weakint                           []                               8
+   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
+   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
+   4 Ops.CONST           :            dtypes.weakint                           []                               2
+   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
+   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
+   7 Ops.CONST           :            dtypes.weakint                           []                               4
+   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
+   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
+  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:2')
   11 Ops.SHRINK          :            dtypes.float                             [10, 6, 8]                       None
   12 Ops.LOAD            :            dtypes.float                             [11]                             None
   13 Ops.STORE           :            dtypes.void                              [9, 12]                          None
   14 Ops.GROUP           :            dtypes.void                              [13, 13, 13, 13]                 None
   15 Ops.SINK            :            dtypes.void                              [14]                             KernelInfo(name='E_8_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(4, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=0, arg=(8, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
 === kernel 21 ===
-   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:2')
-   1 Ops.CONST           :            dtypes.weakint                           []                               8
-   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
-   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
-   4 Ops.CONST           :            dtypes.weakint                           []                               2
-   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
-   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
-   7 Ops.CONST           :            dtypes.weakint                           []                               4
-   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
-   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
-  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:1')
-  11 Ops.SHRINK          :            dtypes.float                             [10, 6, 8]                       None
-  12 Ops.LOAD            :            dtypes.float                             [11]                             None
-  13 Ops.STORE           :            dtypes.void                              [9, 12]                          None
-  14 Ops.GROUP           :            dtypes.void                              [13, 13, 13, 13]                 None
-  15 Ops.SINK            :            dtypes.void                              [14]                             KernelInfo(name='E_8_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(4, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=0, arg=(8, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
-=== kernel 22 ===
-   0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 32, device='CPU:3')
-   1 Ops.CONST           :            dtypes.weakint                           []                               8
-   2 Ops.CAST            :            dtypes.int                               ['8']                            dtypes.int
-   3 Ops.SPECIAL         :            dtypes.int                               [2]                              lidx0
-   4 Ops.CONST           :            dtypes.weakint                           []                               2
-   5 Ops.CAST            :            dtypes.int                               ['2']                            dtypes.int
-   6 Ops.SHL             :            dtypes.int                               [3, 5]                           None
-   7 Ops.CONST           :            dtypes.weakint                           []                               4
-   8 Ops.CAST            :            dtypes.int                               ['4']                            dtypes.int
-   9 Ops.SHRINK          :            dtypes.float                             [0, 6, 8]                        None
-  10 Ops.PARAM           :            dtypes.float                             []                               ParamArg(1, dtypes.float, 32, device='CPU:2')
-  11 Ops.SHRINK          :            dtypes.float                             [10, 6, 8]                       None
-  12 Ops.LOAD            :            dtypes.float                             [11]                             None
-  13 Ops.STORE           :            dtypes.void                              [9, 12]                          None
-  14 Ops.GROUP           :            dtypes.void                              [13, 13, 13, 13]                 None
-  15 Ops.SINK            :            dtypes.void                              [14]                             KernelInfo(name='E_8_4', applied_opts=(Opt(op=OptOps.SPLIT, axis=0, arg=(4, AxisType.UPCAST)), Opt(op=OptOps.SPLIT, axis=0, arg=(8, AxisType.LOCAL))), opts_to_apply=None, estimates=None, beam=0)
-=== kernel 23 ===
    0 Ops.PARAM           :            dtypes.float                             []                               ParamArg(0, dtypes.float, 64, device=('CPU:0', 'CPU:1', 'CPU:2', 'CPU:3'))
    1 Ops.CONST           :            dtypes.weakint                           []                               16
    2 Ops.CAST            :            dtypes.int                               ['16']                           dtypes.int
